test/functional/wallet_reorgsrestore.py

```diff
@@ -90,6 +91,48 @@ def test_coinbase_automatic_abandon_during_startup(self):
         # Verify the coinbase descendant was also marked as abandoned
         assert_equal(wallet0.gettransaction(descendant_tx_id)['details'][0]['abandoned'], True)
 
+    def test_reorg_handling_during_unclean_shutdown(self):
+        self.log.info("Test that wallet doesn't crash due to a duplicate block disconnection event after an unclean shutdown")
+        node = self.nodes[0]
+        # Receive coinbase reward on a new wallet
+        node.createwallet(wallet_name="reorg_crash", load_on_startup=True)
+        wallet = node.get_wallet_rpc("reorg_crash")
+        self.generatetoaddress(node, 1, wallet.getnewaddress(), sync_fun=self.no_op)
+
+        # Restart to ensure node and wallet are flushed
+        self.restart_node(0)
+        wallet = node.get_wallet_rpc("reorg_crash")
+        assert_greater_than(wallet.getwalletinfo()['immature_balance'], 0)
+
+        # Disconnect tip and sync wallet state
+        tip = wallet.getbestblockhash()
+        wallet.invalidateblock(tip)
+        wallet.syncwithvalidationinterfacequeue()
+
+        # Tip was disconnected, ensure coinbase has been abandoned
+        assert_equal(wallet.getwalletinfo()['immature_balance'], 0)
+        coinbase_tx_id = wallet.getblock(tip, verbose=1)["tx"][0]
+        assert_equal(wallet.gettransaction(coinbase_tx_id)['details'][0]['abandoned'], True)
+
+        # Abort process abruptly to mimic an unclean shutdown (no flush)
+        # The best block locator changes might not have been persisted
+        node.process.kill()
+
+        # Start wallet again. Ensure the tx is still abandoned and the wallet has no balance
```

Comment:
"Ensure the tx is still abandoned and the wallet has no balance" sounds too positive to me - the test comment should say clearly that this is a bug in master and the wallet is showing the wrong balance because it incorrectly treats a tx which is in the current chain as abandoned (even if it doesn't crash anymore).

Reply:
Agree, it's a subtle point but important to convey in the comment.
The 2 assertions on line 124 and 125 below are correct individually but incongruent with regards to the overall node state.

Reply:
@mzumsande 
>  tx which is in the current chain
Isn't the tx in a disconnected block?

Reply:
it was in a disconnected node before we killed the node - but the chainstate was not saved to disk with the unclean shutdown, so after the restart that block is part of the chain again - while the wallet (which synced with disk immediately) thinks the tx is not in the chain.
Of course we might do the reorg again if peers give us the better chain again, fixing the problem, but this is not guaranteed: while the node was down, the original chain may have been extended, in which case the balance would be wrong indefinitely.